### PR TITLE
[JENKINS-71366] Set sandbox allow-same-origin in DEFAULT_CSP_VALUE.

### PR DESCRIPTION
**Pull request by mattpulver**
See JENKINS-71366.

The current `DEFAULT_CSP_VALUE` is ineffective w.r.t. the serving of image and stylesheets:
https://github.com/jenkinsci/jenkins/blob/357fde992de740c666a84b9caab520b3301f7204/core/src/main/java/hudson/model/DirectoryBrowserSupport.java#L872

The 2 directives `img-src 'self';` and `style-src 'self';` suggest that the intention is to allow images and stylesheets to be served from the same origin.

However this is disallowed by `sandbox;` without the `allow-same-origin` token. From https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox:

> `allow-same-origin`
> If this token is not used, the resource is treated as being from a special origin that always fails the [same-origin policy](https://developer.mozilla.org/en-US/docs/Glossary/Same-origin_policy) (potentially preventing access to [data storage/cookies](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#cross-origin_data_storage_access) and some JavaScript APIs).

**TL;DR:** Images and stylesheets are not allowed with this `Content-Security-Policy`. Fix it by replacing `sandbox;` with `sandbox allow-same-origin;`.

### Testing done

This was tested by running the following in our Jenkins Script Console:
```
System.setProperty("hudson.model.DirectoryBrowserSupport.CSP", "sandbox allow-same-origin; default-src 'none'; img-src 'self'; style-src 'self';")
```
as described in https://www.jenkins.io/doc/book/security/configuring-content-security-policy/

It was observed that this successfully allowed same-origin CSS files to be served with their HTML files, whereas previously this was prohibited by the default CSP rule.

### Proposed changelog entries

- Add `allow-same-origin` to the `sandbox` Content-Security-Policy directive of workspace and artifact browsers, if the Resource Root URL feature is not used, to allow requests to resources like stylesheets and images, even if a reverse proxy prohibits cross-site requests.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8273"></a>

### CODE REVIEW

**Comment by MarkEWaite** [comment]
Thanks for your first pull request @mattpulver .

Could you restore the pull request template?  It includes items that are mandatory for a Jenkins pull request, including:

* Proposed changelog entry (may be N/A in this case)
* Proposed upgrade guide entry (may be N/A in this case)
* Testing done (description of the interactive testing performed to confirm the change works)

**Comment by mattpulver** [comment]
> > TL;DR: Images and stylesheets are not allowed with this Content-Security-Policy. Fix it by replacing sandbox; with sandbox allow-same-origin;.
> 
> Works for me with basic archived HTML and PNG files locally in both Chrome and Firefox. Is `localhost` special or are there other restrictions for this issue to occur? Comments in the Jira issue are about HTML Publisher and its iframe, but then mention iframes aren't needed…?

Thanks for checking. I was able to reproduce the problem using `localhost`.

Similar to the headers posted in the Jira, would it be possible to share the following Request/Response headers:
1. Request headers for the HTML file (that contains the PNG directly.)
2. Response headers for the HTML.
3. Request headers for the PNG.
4. Response headers for the PNG.

What I'm most interested in is verifying/seeing:
* `Content-Security-Policy: sandbox; default-src 'none'; img-src 'self'; style-src 'self';` is seen in step 2.
* The `Sec-Fetch-Site:` header made by the client in step 3.

The default CSP unintentionally causes `Sec-Fetch-Site: cross-site` to be used/requested by the client in step 3 which in turn causes some web servers to deny the request.

**Comment by daniel-beck** [comment]
> The default CSP unintentionally causes `Sec-Fetch-Site: cross-site` to be used/requested by the client in step 3 which in turn causes some web servers to deny the request.

Thanks, that was what I was missing, as I was running Jenkins not behind nginx. I get the same request headers for the second request, but the second response works, since Jenkins doesn't care about this header.

**Comment by mattpulver** [comment]
That makes sense. With the change made in this PR, the updated CSP causes the client to correctly request
```
Sec-Fetch-Site: same-origin
```
in step 3 instead of `cross-site` (when either an image or stylesheet is requested.)

**Review by daniel-beck — APPROVED** [review]

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
